packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,211 @@
+  | 'adguard_ext_android_cb'
+  | 'adguard_ext_chromium'
+  | 'adguard_ext_edge'
+  | 'adguard_ext_firefox'
+  | 'adguard_ext_opera'
+  | 'adguard_ext_safari'
+  | (string & {});
+
+export class Env extends Map<EnvKeys, boolean> {}
+
+export const enum PreprocessorTypes {
+  INVALID = 0,
+  BEGIF = 1,
+  ELSE = 2,
+  ENDIF = 3,
+}
+
+export function detectPreprocessor(line: string) {
+  if (line.charCodeAt(1) !== 35 /* '#' */) {
+    return PreprocessorTypes.INVALID;
+  }
+
+  const command = line.slice(2);
+
+  if (fastStartsWith(command, 'if ')) {
+    return PreprocessorTypes.BEGIF;
+  }
+
+  if (command === 'else') {
+    return PreprocessorTypes.ELSE;
+  }
+
+  if (command === 'endif') {
+    return PreprocessorTypes.ENDIF;
+  }
+
+  return PreprocessorTypes.INVALID;
+}
+
+const operatorPattern = /(\|\||&&)/g;
+const identifierPattern = /^(!?[a-z0-9_]+)$/;
+
+const tokenize = (expression: string) => expression.split(operatorPattern);
+const matchIdentifier = (identifier: string) => identifier.match(identifierPattern);
+
+export const evaluate = (expression: string, env: Env) => {
+  const tokens = tokenize(expression);
+
+  let result = false;
+  let isContinuedByAndOperator = false;
+
+  for (let i = 0; i < tokens.length; i++) {
+    if (i % 2) {
+      if (tokens[i][0] === '|') {
+        isContinuedByAndOperator = false;
+      } else if (tokens[i][0] === '&') {
+        isContinuedByAndOperator = true;
+      } else {
+        // Invalid expression
+        return false;
+      }
+    } else {
+      const match = matchIdentifier(tokens[i]);
+
+      if (!match) {
+        // Invalid expression
+        return false;
+      }
+
+      let identifier = match[1];
+
+      const isNegated = identifier.charCodeAt(0) === 33; /* '!' */
+      let isPositive = false;
+
+      if (isNegated) {
+        identifier = identifier.slice(1);
+        isPositive = !env.get(identifier);
+      } else {
+        isPositive = !!env.get(identifier);
+      }
+
+      if (isContinuedByAndOperator) {
+        result &&= isPositive;
+      } else {
+        result ||= isPositive;
+      }
+    }
+  }
+
+  return result;
+};
+
+export default class Preprocessor {
+  public static parse(line: string): Preprocessor | null {
+    return new this({
+      condition: line.slice(5 /* '!#if '.length */).replace(/ /g, ''),
+    });
+  }
+
+  public static deserialize(view: StaticDataView): Preprocessor {
+    const condition = view.getUTF8();
+
+    const positives = new Set<number>();
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      positives.add(view.getUint32());
+    }
+
+    const negatives = new Set<number>();
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      negatives.add(view.getUint32());
+    }
+
+    return new this({
+      condition,
+      positives,
+      negatives,
+    });
+  }
+
+  public readonly condition: string;
+
+  public readonly positives: Set<number>;
+  public readonly negatives: Set<number>;
```

Comment:
I would simplify by having each preprocessor block be a single list of IDs. If we want to handle `else` block we can create a second instance of `Preprocessor` with the condition being negated.

Reply:
Do you think introducing `NegatedPreprocessor` (referencing `Preprocessor`) reasonable because we don't implemented grouping via parentheses?

Reply:
This was decided to wrap with parenthesis.*